akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -272,6 +272,24 @@ object GraphStageLogic {
   private[stream] val NoPromise: Promise[Done] = Promise.successful(Done)
 }
 
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ConcurrentAsyncCallbackState {
+  sealed trait State[E]
+  // waiting for materialization completion or during dispatching of initially queued events
+  final case class Pending[E](pendingEvents: List[Event[E]]) extends State[E]
+  // stream is initialized and so no threads can just send events without any synchronization overhead
+  case object Initialized extends State[Any]
+  def initialized[E]: State[E] = Initialized.asInstanceOf[State[E]]
+  // Event with feedback promise
+  final case class Event[E](e: E, handlingPromise: Promise[Done])
+
+  private val NoPendingEvents = Pending(Nil)
+  def noPendingEvents[E]: State[E] = NoPendingEvents.asInstanceOf[State[E]]
+}
```

Comment:
Avoids allocations of the `Initialized` and `NoPending` per `ConcurrentAsyncCallback`, but also gets rid of the implicit `$outer` field in all objects/classes that was there because nested inside a class.

Started out because it showed up in the heap dump, but when it turned out to be a red herring I still thought it may be worth to keep.